node/core/approval-voting/src/lib.rs

```diff
@@ -539,6 +533,95 @@ struct ApprovalStatus {
 	block_tick: Tick,
 }
 
+enum ApprovalOutcome {
+	Approved,
+	Failed(InvalidCandidate),
+}
+
+struct ApprovalState {
+	validator_index: ValidatorIndex,
+	candidate_hash: CandidateHash,
+	candidate_index: CandidateIndex,
+	status: ApprovalOutcome,
+}
+
+impl ApprovalState {
+	fn approved(
+		validator_index: ValidatorIndex,
+		candidate_hash: CandidateHash,
+		candidate_index: CandidateIndex
+	) -> Self {
+		Self {
+			validator_index,
+			candidate_hash,
+			candidate_index,
+			status: ApprovalOutcome::Approved
+		}
+	}
+	fn failed(
+		validator_index: ValidatorIndex,
+		candidate_hash: CandidateHash,
+		candidate_index: CandidateIndex,
+		reason: InvalidCandidate
+	) -> Self {
+		Self {
+			validator_index,
+			candidate_hash,
+			candidate_index,
+			status: ApprovalOutcome::Failed(reason)
+		}
+	}
+}
+
+type BoundedBackgroundApprovalTask = Timeout<RemoteHandle<ApprovalState>>;
+
+struct CurrentlyCheckingSet {
+	approvals_cache: lru::LruCache<CandidateHash, ApprovalOutcome>,
+	candidate_hash_map: HashMap<CandidateHash, Vec<Hash>>,
+	currently_checking: FuturesUnordered<BoundedBackgroundApprovalTask>,
+}
+
+impl Default for CurrentlyCheckingSet {
+	fn default() -> Self {
+		Self {
+			approvals_cache: lru::LruCache::new(APPROVAL_CACHE_SIZE),
+			candidate_hash_map: HashMap::new(),
+			currently_checking: FuturesUnordered::new(),
+		}
+	}
+}
+
+impl CurrentlyCheckingSet {
+	fn insert_relay_block_hash(
+		&mut self,
+		candidate_hash: &CandidateHash,
+		relay_block: Hash,
+	) -> bool {
+		self.candidate_hash_map
+			.get_mut(&candidate_hash)
+			.map(|v| {
+				v.binary_search_by_key(&relay_block, |v| *v)
+				 .map_err(|k| v.insert(k, relay_block))
+				 .map(|_| false)
+				 .unwrap_or(true)
+			})
+			.unwrap_or_default()
+	}
+
+	async fn next(&mut self) -> Option<(Vec<Hash>, ApprovalState)> {
+		self.currently_checking
+			.next()
+			.await
+			.flatten()
```

Comment:
`flatten` here is because the timeout returns `None`? Doesn't that mean that timeouts are ignored and this set will grow indefinitely?

I imagine you're working around the fact that the timeout doesn't yield the `candidate_hash`. The `FuturesUnordered` should probably wrap `Box<dyn Future<Output = ApprovalState>>` so the future can return the candidate hash and an `ApprovalOutcome::Failed` even after the timeout.